contracts/utils/Arrays.sol

```diff
@@ -0,0 +1,43 @@
+pragma solidity ^0.4.23;
+
+import "../math/Math.sol";
+
+
+/**
+ * @title Arrays
+ * @dev Utility library of inline array functions
+ */
+library Arrays {
+
+  /**
+   * @dev Find upper bound for searching element. Utilize binary search function with O(log(n)) cost.
+   */
+  function findUpperBound(
+    uint256[] storage array,
+    uint256 element
+  ) 
+    internal
+    view
+    returns (uint256)
+  {
+    uint256 low = 0;
+    uint256 high = array.length;
+
+    while (low < high) {
+      uint256 mid = Math.average(low, high);
```

Comment:
@nventuro I'm sorry but I really don't see the point here :( Could you provide more context to where it's possible to access array as `array[array.length]' ?

Reply:
It isn't, because of how `Math.average` works, that's my point :grin:

Reply:
Hah.. ok you got me :laughing: So I will try to put some comment here too

Reply:
Or you can propose one.. and make my life better :smiley:

Reply:
Haha sorry! Something like this should work:

```
Note that mid will always be strictly less than high (i.e. it will be a valid array index) because Math.average rounds down (it does integer division with truncation).
```

Reply:
Good thanks! I could never do it better! 😄 I see it's last missing thing for now so I will push it shortly. The last time - @nventuro Thanks a lot for that!